src/Limiters/limiter.jl

```diff
@@ -0,0 +1,116 @@
+"""
+    quasimonotone_limiter!(ρstar, Qstar, minQ, maxQ, space, tol)
+
+This limiter is inspired by Guba et al Optimization-based limiters for the
+spectral element method", JCP 2014, doi: https://doi.org/10.1016/j.jcp.2014.02.029
+In the reference paper, it is denoted by OP1, and is outlined in eq.s (37)-(40).
+As in HOMME, the implementation idea here is the following: We need to find a grid field which is closest
+to the initial field (in terms of weighted sum), but satisfies the min/max constraints.
+So, first we find values which do not satisfy constraints and bring these values
+to a closest constraint. This way we introduce some mass change (`mass_change`),
+so, we redistribute `mass_change` in the way that the l2 error is smallest.
+This redistribution might violate constraints thus, we do a few iterations.
+"""
+function quasimonotone_limiter!(
+    ρstar::Fields.Field,
+    Qstar::Fields.Field,
+    space::Spaces.SpectralElementSpace2D,
+    FT,
+    minQ,
+    maxQ,
+    tol,
+)
+    Nq = Spaces.Quadratures.degrees_of_freedom(Spaces.quadrature_style(space))
+    n_elems = length(minQ)
+    qstar = Array{FT, 4}(undef, Nq, Nq, 1, n_elems)
+    wJ = Array{FT, 2}(undef, Nq, Nq)
+    ρstar_wJ = Array{FT, 2}(undef, Nq, Nq)
+
+    ρstar_e_mass = Array{FT}(undef, n_elems)
+    Qstar_e_mass = Array{FT}(undef, n_elems)
+
+    # Compute ρstar's and Qstar's masses over an element e
+    for e in 1:n_elems
+        ρstar_e_mass[e] = sum(Fields.slab(ρstar, e))
+        Qstar_e_mass[e] = sum(Fields.slab(Qstar, e))
+
+        # This should never happen, but if it does, don't limit
+        if ρstar_e_mass[e] <= 0
+            return Qstar
+        end
+
+        # Relax constraints to ensure limiter has a solution:
+        # This is only needed if runnign with the SSP CFL>1 or
+        # due to roundoff errors.
+        if Qstar_e_mass[e] < minQ[e] * ρstar_e_mass[e]
+            minQ[e] = Qstar_e_mass[e] / ρstar_e_mass[e]
+        end
+
+        if Qstar_e_mass[e] > maxQ[e] * ρstar_e_mass[e]
+            maxQ[e] = Qstar_e_mass[e] / ρstar_e_mass[e]
+        end
+
+        local_geometry_slab = Fields.slab(space.local_geometry, e)
+
+        for iter in 1:(Nq * Nq - 1)
+            mass_change = 0.0
+            # Iterate over quadrature points
+            for i in 1:Nq, j in 1:Nq
+                wJ[i, j] = local_geometry_slab[i, j].WJ
+                ρstar_wJ[i, j] = wJ[i, j] * parent(ρstar)[i, j, 1, e]
```

Comment:
```suggestion
                ρstar_wJ[i, j] = local_geometry_slab[i, j].WJ * parent(ρstar)[i, j, 1, e]
```
and then can get rid of `wJ`